src/sentry/api/utils.py

```diff
@@ -108,6 +115,11 @@ def get_date_range_rollup_from_params(
     interval = int(minimum_interval * math.ceil(interval / minimum_interval))
 
     start, end = get_date_range_from_params(params)
+    # when the `end` is supposed to be inclusive (and only if it was
+    # explicitly given), we simply add a second to it, the rounding logic down
+    # below will take care of the rest.
+    if round_range and inclusive_end and params.get("end"):
+        end += timedelta(seconds=1)
```

Comment:
We also have another place where we add 1 second in [sentry.utils.snuba.SnubaQueryParams](https://github.com/getsentry/sentry/blob/master/src/sentry/utils/snuba.py#L593) but it only applies to the undefined end value.

The comment there makes me think that we might want to consider always adding the 1 second as it mentions the UI wanting to be inclusive but snuba defaults to exclusive.

Reply:
There are a few things at play here:
1. when using `statsPeriod`, 3d/1d should mean 3 slices of 1d. And to avoid artifacts at the borders, we round the start/end to intervals as well, so with 3d, we don’t start at `2021-03-01T08:42:00`, since that way we would lose ~8h of the first bucket.
2. similarly, just adding 1s would maybe work if we have an interval of 1h when querying the hourly table, which works out fine, but when using the raw table, you would only get 1s/60s, which is weird.

That being said, IMO, there has to be some kind of rounding to the interval here.